### PR TITLE
dev-merge-ecg initial AND final redshifts

### PR DESCRIPTION
**Pull request by emikogardiner**
## Description
* Updated cyutils.pyx, single_sources.py, sam.py, and librarian.py to calculate both initial and final redshift for bg and ss, as well as hc^2-weighted final comoving distance, final separation, and final angular separation of the background. Returned in sspar (N,4,F,R,L) and bgpar (N,7,F,R). All pars are in cgs units.
* Added single_sources.all_sspars(fobs_gw_cents, sspar) to calculate sspar in the same format as bgpar, as (7,F,R) for a single sample.
* Added single source detstats to detstats.py and some functions in sam_cython.pyx

### CODE REVIEW

**Comment by emikogardiner** [comment]
Demo in [dev-notebooks/ss_gwb_redz.ipynb](https://github.com/nanograv/holodeck/blob/5fe1f163a9113cac70d457d084bc25f4ffe6d81a/dev-notebooks/ss_gwb_redz.ipynb)

File: holodeck/sam.py [hunk 1]
```diff
@@ -994,13 +994,14 @@ def gwb(self, fobs_gw_edges, hard, realize=100, loudest=1, params=False):
...
-            Returned only if params == True.
-        bgpar : (3, F, R) NDarray of scalars
-            Average effective binary astrophysical parameters for background sources at each frequency 
-            and realization, 
-            Returned only if params == True.
+        sspar : (4, F, R, L) NDarray of scalars
+            Astrophysical parametes (total mass, mass ratio, initial redshift, final redshift) of each 
+            loud single sources, for each frequency and realization. 
+            Returned only if params = True.
+        bgpar : (4, F, R) NDarray of scalars
```
Comment (lzkelley) [inline holodeck/sam.py:1001]: 7?

File: holodeck/utils.py [hunk 2]
```diff
@@ -1782,6 +1782,30 @@ def lambda_factor_dlnf(frst, dfdt, redz, dcom=None):
     return lambda_fact
 
 
+def asep_from_sepa(sepa, dcom, redz):
```
Comment (lzkelley) [inline holodeck/utils.py:1785]: Sure, that's not going to be confusing  😂

File: holodeck/single_sources.py [hunk 3]
```diff
@@ -577,7 +602,7 @@ def ss_by_ndars(edges, number, realize, round = True):
         R=1
         bgnum = np.random.poisson(number[...,np.newaxis], 
                                 size = (number.shape + (R,)))
-        assert (np.all(bgnum%1 ==0)), 'nonzero numbers found with realize=True'
+        assert (np.all(bgnum%1 ==0)), 'non integer numbers found with realize=True'
```
[the remark this answers is not in the document]
Comment (emikogardiner) [reply holodeck/single_sources.py:605]: No, but it's in a function that is never used anymore so I wasn't really trying to improve it, just correcting the typo

File: holodeck/sam_cython.pyx [hunk 4]
```diff
@@ -675,7 +675,7 @@ cdef int _dynamic_binary_number_at_fobs_2pwl(
 @cython.boundscheck(False)
 @cython.wraparound(False)
 @cython.nonecheck(False)
-@cython.cdivision(True)
+@cython.cdivision(False)
```
[the remark this answers is not in the document]
Comment (emikogardiner) [reply holodeck/sam_cython.pyx]: This was unintentional, should be fixed now.

**Pull request merged**